virt/xen/xen.py

```diff
@@ -0,0 +1,294 @@
+#!/usr/bin/env python
+
+import sys
+from time import time
+import XenAPI
+from XenAPI import Failure, NewMaster
+from collections import defaultdict, OrderedDict
+import virt
+
+class Guest(object):
+    """
+    This class represents one virtualization guest running on some
+    host/hypervisor.
+    """
+
+    STATE_UNKNOWN = 0      # unknown state
+    STATE_RUNNING = 1      # running
+    STATE_BLOCKED = 2      # blocked on resource
+    STATE_PAUSED = 3       # paused by user
+    STATE_SHUTINGDOWN = 4  # guest is being shut down
+    STATE_SHUTOFF = 5      # shut off
+    STATE_CRASHED = 6      # crashed
+    STATE_PMSUSPENDED = 7  # suspended by guest power management
+
+    def __init__(self,
+                 uuid,
+                 virt,
+                 state,
+                 hypervisorType=None,
+                 hypervisorVersion=None):
+        """
+        Create new guest instance that will be sent to the subscription manager.
+
+        `uuid` is unique identification of the guest.
+
+        `virt` is a `Virt` class instance that represents virtualization hypervisor
+        that owns the guest.
+
+        `hypervisorType` is additional type of the virtualization, used in libvirt.
+
+        `state` is a number that represents the state of the guest (stopped, running, ...)
+
+        `hypervisorVersion` is the version of the hypervisor software running on
+        the hypervisor, if available. If none is available then the value of
+        this attribute will be an empty string ""
+            - This attribute is only included in the dictionary representation
+              if the hypervisorType is not None.
+        """
+        self.uuid = uuid
+        self.virtWhoType = virt.CONFIG_TYPE
+        self.hypervisorType = hypervisorType
+        self.state = state
+        self.hypervisorVersion = hypervisorVersion or ""
+
+    def toDict(self):
+        d = dict((
+            ('guestId', self.uuid),
+            ('state', self.state),
+            ('attributes', {
+                'virtWhoType': self.virtWhoType,
+                'active': 1 if self.state in (self.STATE_RUNNING, self.STATE_PAUSED) else 0
+            }),
+        ))
+
+        if self.hypervisorType is not None:
+            d['attributes']['hypervisorType'] = self.hypervisorType
+            if self.hypervisorVersion is not None or \
+               self.hypervisorVersion is not "":
+                d['attributes']['hypervisorVersion'] = self.hypervisorVersion
+        return d
+
+class Hypervisor(object):
```

Comment:
I have worked based on version 14.1. I *think* (out of my memory) these kind of things changed in version 15. I'll take a look, but I may need some mentoring